feat(tooltip): add definition tooltip

aledavila opened the pull request:
## Overview

https://github.ibm.com/carbon/issues/issues/618
https://github.ibm.com/carbon/issues/issues/549#issuecomment-4988198 
https://github.com/carbon-design-system/carbon-components/issues/771


### Added

Added new tooltips with on click tooltip being the default.
Added definition tooltip
Added icon tooltip

### Removed

Removed on hover tooltip.
Removed simple tooltip

### Changed

Simple tooltip is just for defining the icon.
Default font weight is now normal. Bold is a modifier. 


## Notes

Tayler needs to sign of and look at styles.

aledavila committed:
feat(tooltip): add definition tooltip

aledavila committed:
feat(tooltip): add on click tooltip

aledavila committed:
chore(tooltip): simple info tooltip

alisonjoseph commented:
@aledavila These don't look like they are related to this PR?  #618 #549

alisonjoseph commented on src/components/tooltip/_tooltip.scss:
```diff
@@ -140,13 +142,12 @@
-      max-width: rem(240px);
-      padding: $spacing-lg;
+      max-width: rem(170px);
+      padding: rem(10px) $spacing-xs;
```
Might be a question for design but can `rem(10px)` be tweaked up or down to fit into our spacing scale?

aagonzales replied:
It should be 8px

aledavila commented:
@alisonjoseph yes you are right sorry. those are the issue #s in IBM carbon repo no this repo

aledavila commented:
Fixed the links

alisonjoseph reviewed (changes requested):
Overall looks good 😄 

Looks like the triangle needs to move 1px down? Screen shot of a screen shot I zoomed in on. 
<img width="520" alt="screen shot 2018-04-19 at 12 40 28 pm" src="https://user-images.githubusercontent.com/2753488/39008665-ee197806-43ce-11e8-8ad9-9d002dfbfb85.png">

Changing the functionality of the existing tooltip would be a breaking change. (from hover to click). Would maybe be better to add it as a new variant or v2, thoughts?

aledavila commented:
Thanks I’ll look into the changes. As it stands how the changes were made it doesn’t break anything rather people who have the tooltip implemented would have to click on it now instead of hovering when they pull the new changes. But i don’t know how you guys prefer to handle that. 

Yea I’ll ask Tayler about the changes probably better to have the padding all spacing-xs.

asudoh commented:
Agreed with @alisonjoseph that looks great overall 🎉 - Thanks @aledavila!

The only question I have is how to handle the change in behavior, which would need to involve designers (CC @carbon-design-system/design) - Should we introduce some way to keep the older behavior, or should we try something to ensure that the change in behavior does not come with a surprise to our consuming services...?

aledavila committed:
fix(tooltip): fix arrowhead 1 pixel

aledavila commented:
@alisonjoseph  I changed the 1 pixel that was off. I'll ask Tayler about the other padding issue when she gets back.

aagonzales commented:
> Changing the functionality of the existing tooltip would be a breaking change. (from hover to click). Would maybe be better to add it as a new variant or v2, thoughts?

I think making it a v2 and deprecating the v1 version with Carbon V10 would be fine. Consumers just won't ever automatically get those update, correct? They'll have to manually update it to v2?

alisonjoseph commented:
correct @aagonzales. That's how we've been handling other components with breaking changes.

aledavila commented:
Oh I see. Can someone explain how that process is or where to make it a V2 instead?

alisonjoseph commented:
Looks like the process for the other components is just updating the classname, so `bx--tooltip-v2` and just make sure that the original tooltip doesn't change. You can see how the v2 input styles are handled in this PR for an example. https://github.com/carbon-design-system/carbon-components/pull/712

aledavila committed:
chore(tooltip): add v2 of tooltip in seperate folder

aledavila committed:
chore(tooltip): add previous version of tooltip

aledavila committed:
chore(tooltip): add files to globals

aledavila commented:
@alisonjoseph I added the v2 now. I just need to ask Tayler about the padding on the simple tooltip. Everything else should be good.

alisonjoseph commented on src/components/tooltip-v2/tooltip-v2--simple.html:
```diff
@@ -0,0 +1,8 @@
+<div class="bx--tooltip--simple-v2">
```
Should this class be `bx--tooltip-v2--simple`?

asudoh commented on src/components/tooltip-v2/migrate-to-7.x.md:
```diff
@@ -0,0 +1,25 @@
+### HTML
```
Given V2 is newly introduced component, does this migration guide still make sense? One thing that comes to my mind is V1 to V2 migration guide. Another approach may be simply keeping the README in-sync with what V2 does.

asudoh reviewed (commented):
Hi @aledavila great to see you complete V2 split!

tay-aitken commented:
@aledavila can you create a staging link so that I can inspect the element and see how everything looks. Thanks! This will help me make a decision about the padding.

aledavila committed:
fix(tooltip): fix simple tooltip class name

aledavila committed:
chore(tooltip): add definition tooltip to readme

tay-aitken commented:
### Fixes
Please take a a look at the [specs](https://github.ibm.com/carbon/issues/issues/549#issuecomment-5477046) I posted previously to make sure the coded variation matches the designs and make the following changes below:

- [ ] Change default tooltip name to Interactive
- [x] Change simple name to Icon

**Interactive tooltip**
- [x] Using the incorrect icon. Let me know if you need an SVG of the right one/if it needs to be added to our icon list. The SVG can be found [here](https://ibm.box.com/s/2wew5q4lzrfocdzcsomcv6vxybmll6bb)
- [x] add in a link and button as seen [here](https://github.ibm.com/carbon/issues/issues/549#issuecomment-5477046)
- [x]  Tooltip should not close if i click inside the tooltip, should only close if I click outside of the tooltip area
- [x] background is not black, it should be `background-color: #272D33;` (and should be using our color tokens)
- [x] Tooltip icon should be darkening on hover by 10% not lightening (sync with TJ, his PR might already be taking care of this)
- [x] Tooltip label should not be changing color on hover, should remain #152935
- [x] Can't inspect the tooltip box itself, having trouble locating it in the code, will review colors, type sizes, padding, etc. once link and button are added in

**Definition Tooltip**
- [x] Please make the tooltip have the following content "Brief description of the dotted, underlined word above."
- [x]  Is there a reason the old CSS for the tooltip still is a part of this component?
- [x] Padding on all sides should be .05rem
- [x] Dotted underline should be #8897A2 (ui-04) on enabled
- [x]  on :hover it should be #3d70b2 (brand-01)
- [x] should be `font-size: 0.75rem;` not 0.875rem
- [x] Tooltip icon should be darkening on hover by 10% not lightening (sync with TJ, his PR might already be taking care of this)
- [x] Double check that the tooltip is only 170px wide, perhaps add in a max-width
- [x] Change label to Definition label, instead of Tooltip label

**Icon Tooltip**
- [x] Should be `background-color: #272D33;`, not white
- [x] Should be `border: none;`
- [x] Remove the label
- [x] Use Filter icon instead of the tooltip icon
- [x] Tooltip content should say "Filter"
- [x] Tooltip is not aligned with top of icon
- [x] Padding should be 4px (.025rem) on all sides
- [x] Should be `font-size: 0.75rem;` (12px)
- [x] Make sure the shark fin is behind the tooltip background
- [x] For example purposes please have the tooltip appear below the icon, not above.

aledavila committed:
chore(tooltip): add tooltip icon

aledavila committed:
chore(tooltip): change definition tooltip to seperate element

aledavila commented:
@alisonjoseph 

Ignore the colors. They will be updated once TJs updated colors are merged. The CSS tooltips have the right color but as a hex for the time being. Renaming Default Tooltip to Interactive Tooltip is more complex than simply changing file name. Possible workaround could be name change in website itself. 

Definition: I changed the definition tooltip completely and took made it a separate element instead of creating it with a before. It helped with being able to give it a max width and if more than the max widt
h it will wrap appropriately. And re did the CSS for it. 

Default: As for the JS I added the prevent close when clicking inside the tooltip. Only close when clicking outside. 

Simple --> Icon: Simple tooltip is now Icon only tooltip.

marijohannessen reviewed (approved):
Looks good to me! 🎉 Would it make sense to lose the `v2` part maybe and target this PR towards the `v9` branch instead so we can release it with the new inputs etc?

aledavila commented:
Yea that sounds good. @marijohannessen would that mean just removing the default and just changing V2 naming back to its default so it’s the default tooltip ? Or Should I pull V9 and make a new branch and just copy the updates onto there ?

marijohannessen commented:
Yeah, just remove `v2` from all classes etc. And then you can just point this PR to the `v9` branch instead of `master`.

If you click on "Edit" next to the PR title:

<img width="1035" alt="screen shot 2018-05-04 at 3 25 08 pm" src="https://user-images.githubusercontent.com/5447411/39650880-6f29cdba-4faf-11e8-8e83-afe853c445c4.png">

and then change the base from master to v9:

<img width="947" alt="screen shot 2018-05-04 at 3 25 14 pm" src="https://user-images.githubusercontent.com/5447411/39650875-6b8c23ec-4faf-11e8-8056-2e476d4df4dd.png">

aledavila committed:
chore(tooltip): fix merge conflicts

aledavila committed:
fix(tooltip): fix merge conflict

aledavila committed:
fix(tooltip): fix merge conflict again

aledavila committed:
chore(tooltip): removed v2 and added new color variables

aledavila committed:
chore(tooltip): add bold modifier

aledavila committed:
chore(tooltip): updated migration to v9 and readme files

alisonjoseph reviewed (approved):
LGTM! 👍

asudoh reviewed (commented):
LGTM basically - Great to see this PR getting there @aledavila! Just one little thing - Is it possible for you to remove `yarn-error.log` and `package-lock.json` from the source control before merging? Thanks!

aledavila committed:
chore(tooltip): add focus to definition

aledavila committed:
chore(tooltip): remove tab

alisonjoseph commented:
@aledavila since the demo template changes have been merged in to the v9 branch, would be nice to update this from html to handlebar templates before we merge it in.

asudoh commented:
^ Yes it'll be nice - Simple rename from `.html` to `.hbs` will work for now, but great to merge several component variant's markup using the power of Handlebars. Feel free to ping me if you have any questions - Thanks!

alisonjoseph reviewed (changes requested):
Would be good to add correct aria roles.  Ref https://github.com/carbon-design-system/carbon-components/issues/771

https://www.w3.org/TR/wai-aria-practices/#tooltip

>WAI-ARIA Roles, States, and Properties
The element that serves as the tooltip container has role tooltip.
>The element that triggers the tooltip references the tooltip element with aria-describedby.

tw15egan commented on src/components/tooltip/_tooltip.scss:
```diff
@@ -20,43 +20,64 @@
+    margin-top: $spacing-xs;
+    padding: $spacing-md;
     border: 1px solid $ui-03;
+    border-radius: 4px;
```
Can we switch this to `rem(4px)`?

tw15egan commented on src/components/tooltip/_tooltip.scss:
```diff
@@ -112,20 +133,99 @@
+    margin-top: $spacing-sm;
+    padding: $spacing-xs;
+    color: $inverse-01;
+    border-radius: 4px;
```
Can we switch this to `rem(4px)`?

tw15egan commented on src/components/tooltip/_tooltip.scss:
```diff
@@ -112,20 +133,99 @@
+      left: 0;
+      width: 0.6rem;
+      height: 0.6rem;
+      margin-left: 24px;
```
Can we switch this to `rem(24px)`?

tw15egan commented on src/components/tooltip/_tooltip.scss:
```diff
@@ -134,17 +234,18 @@
       margin-left: 50%;
-      color: $text-01;
+      padding: $spacing-2xs;
+      border-radius: 4px;
```
Same as above

aledavila committed:
chore(tooltip): fix merge conflict

tw15egan reviewed (approved):
Just a few tiny changes, but other than that it looks great!

aledavila commented:
WIP: working on accessibility

aledavila committed:
fix(tooltip): added accessibility aria tags

aledavila committed:
chore(tooltip): change arrow to be a dom element

alisonjoseph commented on src/components/tooltip/tooltip.hbs:
```diff
@@ -1,10 +1,18 @@
-    <path d="M8 0C3.6 0 0 3.6 0 8s3.6 8 8 8 8-3.6 8-8-3.6-8-8-8zm0 4c.6 0 1 .4 1 1s-.4 1-1 1-1-.4-1-1 .4-1 1-1zm2 8H6v-1h1V8H6V7h3v4h1v1z"></path>
-  </svg>
-</a>
+<div class="bx--tooltip__label">
```
We should add the aria-describedby here `<div class="bx--tooltip__label" aria-describedby="unique-tooltip">`

alisonjoseph commented on src/components/tooltip/tooltip.hbs:
```diff
@@ -1,10 +1,18 @@
-</a>
+<div class="bx--tooltip__label">
+  Tooltip Label
+  <div tabindex="0" data-tooltip-trigger data-tooltip-target="#unique-tooltip" role="tooltip" aria-labelledby="unique-tooltip" class="bx--tooltip__trigger">
```
you can remove `aria-labelledby="unique-tooltip"` here

aledavila committed:
chore(tooltip): change aria label location

aledavila committed:
chore(tooltip): attribute name change

alisonjoseph reviewed (approved):
Looks good! 🎉
